src/mind/neuron.rs

```diff
@@ -2,53 +2,65 @@ use std::fmt;
 
 use rand::Rng;
 
-use crate::engine::value::{add, mul, tanh, Node};
+use crate::engine::{
+    node::{add, checkpoint, mul, tanh, Node},
+    state::State,
+};
 
 pub struct Neuron {
-    weights: Vec<usize>,
-    bias: usize,
+    pub name: String,
+    pub weights: Vec<usize>,
+    pub bias: usize,
 }
 
 impl Neuron {
-    pub fn new(n_in: u32, nodes: &mut Vec<Node>) -> Neuron {
+    pub fn new(n_in: u32, name: String, state: &mut State) -> Neuron {
         let mut rng = rand::thread_rng();
 
         let mut weights = vec![];
         for n in 1..=n_in {
-            let w_idx = Node::new(rng.gen_range(-1.0..1.0), format!("w{}", n), nodes);
+            let w_idx = Node::new(
+                rng.gen_range(-1.0..1.0),
+                format!("{}w{}", name, n),
+                true,
+                state,
+            );
             weights.push(w_idx);
         }
 
-        let bias = Node::new(rng.gen_range(-1.0..1.0), String::from("b"), nodes);
+        let bias = Node::new(rng.gen_range(-1.0..1.0), format!("{}b", name), true, state);
 
-        Neuron { weights, bias }
-    }
-
-    pub fn parameters(&self) -> Vec<usize> {
-        let mut params = self.weights.clone();
-        params.push(self.bias);
-        return params;
+        Neuron {
+            name,
+            weights,
+            bias,
+        }
     }
-}
 
-pub fn apply(neuron: &Neuron, xs: &Vec<usize>, nodes: &mut Vec<Node>) -> usize {
-    let mut act: Option<usize> = None;
-    for i in 0..neuron.weights.len() {
-        match act {
-            Some(v) => {
-                act = Some(add(v, mul(neuron.weights[i], xs[i], nodes), nodes));
-            }
-            None => {
-                act = Some(mul(neuron.weights[i], xs[i], nodes));
+    pub fn apply(&self, xs: &Vec<usize>, state: &mut State) -> usize {
+        let mut act: Option<usize> = None;
+        for (i, w) in self.weights.iter().enumerate() {
+            let mul_res = mul(*w, xs[i], state);
+            match act {
+                Some(v) => {
+                    act = Some(add(v, mul_res, state));
+                }
+                None => {
+                    act = Some(mul_res);
+                }
             }
         }
-    }
 
-    tanh(add(act.unwrap(), neuron.bias, nodes), nodes)
+        checkpoint(
```

Comment:
wrapping a neuron's apply in a checkpoint prevents the neuron from being affected before all upstream back propagation is finished and back is called directly on this neurons checkpoint.